### PR TITLE
b-3-Admin

### PR DESCRIPTION
**Pull request by hiixdayah**
## Description:
-complete use case got getStudents, getSenseis, register, update admin, get admin 
## Checklist:
- [x] Merged latest develop
- [x] PR title makes sense

### CODE REVIEW

**Review by kengthong — COMMENTED** [review]
The updateAdmin logic seems to be wrong.
shld be:
if (
      user.accountId != accountId &&
      user.permissionType != ADMIN_PERMISSION_ENUM_OPTIONS.SUPERADMIN
    ) {
    
    GetAdmin shld have jwt authentication checks + isAdmin middleware. Check the teams backend channel to check the middleware part.

Naming conventions for a few endpoints could be improved.
Could add jwt authentication to some endpoints

File: src/config/auth/passport.ts [hunk 1]
```diff
@@ -105,4 +106,28 @@ module.exports = function (passport) {
...
   });
+
+  passport.use(
+    'admin-local',
+    new LocalStrategy({ usernameField: 'email' }, (email, password, done) => {
+      // Match user
+      Admin.findOne({
+        where: { email },
+      }).then((student) => {
+        if (!student) {
```
Comment (ianleck) [inline src/config/auth/passport.ts]: admin

File: src/config/auth/passport.ts [hunk 2]
```diff
@@ -105,4 +106,28 @@ module.exports = function (passport) {
...
+          return done(null, false, { message: 'Invalid email or password' });
+        }
+
+        // Match password
+        bcrypt.compare(password, student.password, (err, isMatch) => {
```
Comment (ianleck) [inline src/config/auth/passport.ts]: admin

File: src/config/auth/passport.ts [hunk 3]
```diff
@@ -105,4 +106,28 @@ module.exports = function (passport) {
...
+          if (err) throw err;
+          if (isMatch) {
+            return done(null, student);
```
Comment (ianleck) [inline src/config/auth/passport.ts]: admin

File: src/services/admin.service.ts [hunk 4]
```diff
@@ -34,46 +38,77 @@ export default class AdminService {
...
         accountId: Utility.generateUUID(),
         username,
         email,
         password,
         userType: USER_TYPE_ENUM_OPTIONS.ADMIN,
         permission: ADMIN_PERMISSION_ENUM_OPTIONS.ADMIN,
-        updatedBy: admin,
-        createdBy: admin,
+        updatedBy: adminCreator,
+        createdBy: adminCreator,
```
Comment (ianleck) [inline src/services/admin.service.ts:56]: Should it be an object? Or will just the ID suffice?

File: src/controllers/admin.controller.ts [hunk 5]
```diff
@@ -20,26 +28,100 @@ export class AdminController {
   }
 
   public static async updateAdmin(req, res) {
+    const { user } = req; //user is the user who is making the request
+    const { accountId } = req.params; //accountId of the admin who is being updatred
     const { admin } = req.body;
+
+    //check if superadmin is making the update
+    if (
+      user.accountId == accountId &&
```
Comment (ianleck) [inline src/controllers/admin.controller.ts]: !=

if you are not user && not superadmin, will send error
if you are user && not superadmin, will not send error
if you are not user && superadmin, will not send error

File: src/services/admin.service.ts [hunk 6]
```diff
@@ -34,46 +38,77 @@ export default class AdminService {
...
+  public static async updateAdmin(accountId: string, adminAccount) {
+    const admin = await Admin.findByPk(accountId);
     if (admin) {
       await admin.update({
-        username: adminDto.username,
-        email: adminDto.email,
+        firstName: adminAccount.firstName,
+        lastName: adminAccount.lastName,
+        contactNumber: adminAccount.contactNumber,
+        permission: adminAccount.permission,
```
Comment (ianleck) [inline src/services/admin.service.ts]: This should be a separate endpoint to set role

File: src/services/admin.service.ts [hunk 7]
```diff
@@ -34,46 +38,77 @@ export default class AdminService {
...
+
+  public static async getAllStudents() {
+    const students = Student.findAll();
+    return students;
+  }
+
+  public static async getAllSenseis() {
+    const senseis = Sensei.findAll();
+    return senseis;
+  }
```
Comment (ianleck) [inline src/services/admin.service.ts]: where: { status: { [Op.ne]: STATUS.BANNED } }

File: src/services/admin.service.ts [hunk 8]
```diff
@@ -34,46 +38,77 @@ export default class AdminService {
...
+
+  public static async findAdminByIdAndRemove(adminId: string) {
+    let admin;
+    try {
+      admin = await Admin.findOne({ where: { adminId } });
+      return Admin.destroy({ where: { adminId } });
+    } catch (e) {
+      throw new Error(ERRORS.ADMIN_DOES_NOT_EXIST);
+    }
+  }
```
Comment (ianleck) [inline src/services/admin.service.ts]: If not done, comment it out

File: src/routes/schema/admin.schema.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  }),
+
+  updateAdmin: joi.object({
+    admin: joi.object({
+      firstName: joi.string(),
+      lastName: joi.string(),
+      contactNumber: joi.number(),
+      permission: joi
+        .string()
+        .valid(...Object.values(ADMIN_PERMISSION_ENUM_OPTIONS)),
```
Comment (ianleck) [inline src/routes/schema/admin.schema.ts:47]: No longer relevant

File: src/routes/schema/admin.schema.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    }),
+  }),
+
+  getAdmin: joi.object({
+    accountId: joi.string().required(),
+  }),
```
Comment (ianleck) [inline src/routes/schema/admin.schema.ts]: duplicate of `adminIdQ`

File: src/routes/admin.route.ts [hunk 11]
```diff
@@ -25,28 +25,42 @@ import admin from './schema/admin.schema';
...
 const router = express.Router();
-
+const passport = require('passport');
 const schemaValidator = require('express-joi-validation').createValidator({});
 
+//create admin account
+//the schema must have details needed to register an admin
 router.post(
   '/register-admin',
   schemaValidator.body(admin.registerAdmin),
```
Comment (ianleck) [inline src/routes/admin.route.ts:36]: Add passport.authenticate('isAuthenticated', { session: false }),

File: src/routes/admin.route.ts [hunk 12]
```diff
@@ -25,28 +25,42 @@ import admin from './schema/admin.schema';
...
 
-//get admin details by admin ID
-router.get(
-  '/view-admin/:id',
-  Utility.asyncHandler(AdminController.viewAdminDetails)
+//get list of senseis
+router.get('/senseis', Utility.asyncHandler(AdminController.getSenseis));
+
+router.delete(
+  '/delete-admin/:id',
```
Comment (ianleck) [inline src/routes/admin.route.ts]: don't need the `/delete-admin` unless you wanna delete multiple types from admin endpoint

File: src/routes/admin.route.ts [hunk 13]
```diff
@@ -25,28 +25,42 @@ import admin from './schema/admin.schema';
...
+  '/',
+  passport.authenticate('isAuthenticated', { session: false }),
+  schemaValidator.params(admin.adminIdQ),
   schemaValidator.body(admin.updateAdmin),
   Utility.asyncHandler(AdminController.updateAdmin)
+); //body are details of admin to be updated without accountId inside, adminIdQ is the id of the USER requesting this
+
+//get admin
+router.get(
+  '/get-admin/:id',
```
Comment (ianleck) [inline src/routes/admin.route.ts]: don't need `/get-admin`

File: src/controllers/admin.controller.ts [hunk 14]
```diff
@@ -20,26 +28,100 @@ export class AdminController {
...
+    try {
+      const admin = await AdminService.findAdminByIdAndRemove(adminId);
+      return apiResponse.result(
+        res,
+        { message: 'Successfullly deleted' },
+        httpStatusCodes.OK
+      );
+    } catch (e) {
+      logger.error('[adminController.deleteAdmin]:' + e.toString());
+      return apiResponse.error(res, 400, { message: e.toString() });
```
Comment (ianleck) [inline src/controllers/admin.controller.ts]: `httpStatusCodes.BAD_REQUEST`

File: src/controllers/admin.controller.ts [hunk 15]
```diff
@@ -20,26 +28,100 @@ export class AdminController {
...
-    //   const { admin } = req.body;
-    //  try {
+  public static async getSenseis(req, res) {
+    const type = req.body;
+    try {
+      const senseis = await AdminService.getAllSenseis(type);
+      return apiResponse.result(res, senseis, httpStatusCodes.OK);
+    } catch (e) {
+      logger.error('[adminController.getUsers]:' + e.toString());
+      return apiResponse.error(res, 400, { message: e.toString() });
```
Comment (ianleck) [inline src/controllers/admin.controller.ts]: `httpStatusCodes.BAD_REQUEST`

File: src/controllers/admin.controller.ts [hunk 16]
```diff
@@ -20,26 +28,100 @@ export class AdminController {
...
       logger.error('[adminController.updateAdmin]' + e.toString());
+      return apiResponse.error(res, httpStatusCodes.BAD_REQUEST, {
+        message: e.toString(),
+      });
+    }
+  }
+
+  public static async getAdmin(req, res) {
+    const { adminId } = req.body;
+    try {
```
Comment (ianleck) [inline src/controllers/admin.controller.ts:101]: do the check for permission as above

**Review by ianleck — CHANGES_REQUESTED** [review]

**Pull request merged**